build_cffi.py

```diff
@@ -60,29 +66,44 @@
 
 
 # =================================================================
+# NOTES FOR DEVELOPERS:
+#   The CFFI implementation works as follows:
+#       - All function prototypes, global variables and type definitions *directly* used
+#           in the python wrapper must be declared via ffi.cdef("""C CODE""").
+#           There must be no compiler directives in this code (#include, #define, etc)
+#       - All implementations of global variables and types present in the cdef() calls
+#           must also be present in the second argument of set_source.
+#           This is passed to the compiler.
+#       - The `sources` kwarg then contains all the .c files in the library which are to be compiled
 
 # This is the overall C code.
 ffi.set_source(
     "py21cmfast.c_21cmfast",  # Name/Location of shared library module
     """
-    #define LOG_LEVEL {log_level}
-
-    #include "GenerateICs.c"
-    """.format(
-        log_level=log_level
-    ),
+    #include "21cmFAST.h"
+    """,
+    sources=c_files,
     include_dirs=include_dirs,
     library_dirs=library_dirs,
     libraries=libraries,
     extra_compile_args=extra_compile_args,
 )
 
-# This is the Header file
-with open(os.path.join(CLOC, "21cmFAST.h")) as f:
+# Header files containing types, globals and function prototypes
```

Comment:
If we have a format for the wrapper files, could we just do:

```python
wrapper_files = Path(CLOC).glob("_*_wrapper.h")
for fl in wrapper_files:
    with open(fl, 'r') as f:
        ffi.cdef(f)
```

Reply:
I'm not sure about this, I think it might be a good idea to strictly keep these wrapper file categories as input parameters, output structures and function prototypes, with the possible addition of global variables (although we would want to minimize these). If there is a use-case for exposing more than this we can change it but I think we should be keeping the entry/exit points to the backend minimal.

Reply:
Yeah, you make a good point. For a small number like this, explicit is better than implicit.